File: ruby/src/lib/format.rb
# frozen_string_literal: true
# rbs_inline: enabled

require 'csv'
require 'json'

module Lib
  # Converts a CSV of question/answer pairs into the JSON shape Botpress
  # expects for Q&A training data.
  module Format
    # @rbs array: Array[untyped]
    # @rbs return: Hash[Symbol, untyped]
    def template(array: [])
      {
        id: '',
        data: {
          action: 'text', contexts: ['sample'], enabled: true,
          answers: { ja: array }, questions: { ja: array.dup },
          redirectFlow: '', redirectNode: ''
        }
      }
    end

    # @rbs path_to_csv_training_data: String
    # @rbs array: Array[untyped]
    # @rbs return: String
    def to_json(path_to_csv_training_data:, array: [])
      result = array
      format = template(array: array.dup)
      CSV.foreach(path_to_csv_training_data, headers: true) do |training_datum|
        format = merge_training_datum(format, training_datum)
        format[:data][:questions][:ja].uniq!
        result << format
      end
      JSON.pretty_generate({ qnas: result.uniq })
    end

    private

    # @rbs format: Hash[Symbol, untyped]
    # @rbs training_datum: CSV::Row
    # @rbs return: Hash[Symbol, untyped]
    def merge_training_datum(format, training_datum)
      if format[:data][:answers][:ja].last == training_datum['Answer']
        format[:data][:questions][:ja] << training_datum['Question']
        format
      else
        start_new_format(training_datum)
      end
    end

    # @rbs training_datum: CSV::Row
    # @rbs return: Hash[Symbol, untyped]
    def start_new_format(training_datum)
      new_format      = template
      new_format[:id] = training_datum['ID']
      new_format[:data][:questions][:ja] << training_datum['Question']
      new_format[:data][:answers][:ja] << training_datum['Answer']
      new_format
    end
  end
end

File: python/src/lib/format.py
import csv
import json
from typing import Any


class Format:
    """Converts a CSV of question/answer pairs into the JSON shape Botpress
    expects for Q&A training data. Designed to be mixed into a class that
    owns the CSV path."""

    def template(self, array: list[Any] | None = None) -> dict[str, Any]:
        items = list(array) if array is not None else []
        return {
            'id': '',
            'data': {
                'action': 'text',
                'contexts': ['sample'],
                'enabled': True,
                'answers': {'ja': items},
                'questions': {'ja': list(items)},
                'redirectFlow': '',
                'redirectNode': '',
            },
        }

    def to_json(self, path_to_csv_training_data: str,
                array: list[Any] | None = None) -> str:
        result: list[dict[str, Any]] = list(array) if array is not None else []
        fmt = self.template(array=list(array) if array is not None else [])
        with open(path_to_csv_training_data) as f:
            for training_datum in csv.DictReader(f):
                fmt = self._merge_training_datum(fmt, training_datum)
                fmt['data']['questions']['ja'] = self._uniq(
                    fmt['data']['questions']['ja'])
                result.append(fmt)
        return json.dumps({'qnas': self._uniq(result)},
                          ensure_ascii=False, indent=2)

    # private

    def _merge_training_datum(
            self, fmt: dict[str, Any],
            training_datum: dict[str, str]) -> dict[str, Any]:
        answers = fmt['data']['answers']['ja']
        if answers and answers[-1] == training_datum['Answer']:
            fmt['data']['questions']['ja'].append(training_datum['Question'])
            return fmt
        return self._start_new_format(training_datum)

    def _start_new_format(
            self, training_datum: dict[str, str]) -> dict[str, Any]:
        new_format = self.template()
        new_format['id'] = training_datum['ID']
        new_format['data']['questions']['ja'].append(training_datum['Question'])
        new_format['data']['answers']['ja'].append(training_datum['Answer'])
        return new_format

    @staticmethod
    def _uniq(items: list[Any]) -> list[Any]:
        result: list[Any] = []
        for item in items:
            if item not in result:
                result.append(item)
        return result
